v2/tools/generator/internal/codegen/pipeline/state.go

```diff
@@ -51,10 +57,31 @@ func (s *State) WithConversionGraph(graph *storage.ConversionGraph) *State {
 		panic("may only set the conversion graph once")
 	}
 
-	return &State{
-		definitions:     s.definitions.Copy(),
-		conversionGraph: graph,
+	result := s.copy()
+	result.conversionGraph = graph
+	return result
+}
+
+// WithSeenStage records that the passed stage has been seen
+func (s *State) WithSeenStage(id string) *State {
+	result := s.copy()
+	result.stagesSeen.Add(id)         // Record that we saw this stage
+	delete(result.stagesExpected, id) // Discard expectations as they are satisfied
+	return result
+}
+
+// WithExpectation records our expectation that the later stage is coming
+func (s *State) WithExpectation(earlierStage string, laterStage string) *State {
+	result := s.copy()
```

Comment:
Same as above!

Reply:
As discussed above - `State` is an immutable object, we always want to get a new copy.